### DISCUSSION TITLE
maybe a bug in ecole.observation.Pseudocosts?

### DISCUSSION THREAD

**Discussion by ghost** [ds4dm/ecole#307] [Questions] [answered]
when I try to run the example as follows:
```python
class ExploreThenStrongBranch:
    def __init__(self, expert_probability):
        self.expert_probability = expert_probability
        self.pseudocosts_function = ecole.observation.Pseudocosts()
        self.strong_branching_function = ecole.observation.StrongBranchingScores()

    def before_reset(self, model):
        self.pseudocosts_function.before_reset(model)
        self.strong_branching_function.before_reset(model)

    def extract(self, model, done):
        probabilities = [1-self.expert_probability, self.expert_probability]
        expert_chosen = bool(np.random.choice(np.arange(2), p=probabilities))
        if expert_chosen:
            return (self.strong_branching_function.extract(model,done), True)
        else:
            return (self.pseudocosts_function.extract(model,done), False)
        
        
scip_parameters = {'separating/maxrounds': 0, 'presolving/maxrestarts': 0,
                   'limits/time': 100, 'timing/clocktype': 2}
observation_function = { "scores": ExploreThenStrongBranch(expert_probability=0),
                         "node_observation": ecole.observation.NodeBipartite() }
env = ecole.environment.Branching(observation_function=observation_function,
                                  scip_params=scip_parameters, pseudo_candidates=True)


env.seed(0)
instance_path = 'data/instances/setcover/train_500r_1000c_0.05d/instance_0000.lp'
instance = ecole.scip.Model.from_file(instance_path)
observation, action_set, _, done, _ = env.reset(instance)
```

I find something hard to understand:
(1) 
input `observation['scores'][0].shape`
output `994`
not equal to num of variables 1000, why?
(2) 
input `observation['scores'][0][action_set]`
output 
```
array([       nan,        nan,        nan,        nan,        nan,
              nan,        nan,        nan,        nan,        nan,
              nan,        nan,        nan,        nan,        nan,
              nan, 0.18499214, 0.05232925, 0.11698202,        nan,
       0.18961997,        nan, 0.10188215, 0.17122907, 0.18852313,
              nan, 0.17757535, 0.19141571, 0.20885814, 0.20852026,
       0.11995715, 0.11698202, 0.15915762, 0.20743212, 0.06429137,
       0.1105206 , 0.17639468, 0.20891465, 0.05871806,        nan,
       0.20902973, 0.16134182, 0.04683438, 0.10060939, 0.04683438,
       0.03530609, 0.19141571, 0.17757535, 0.14443708, 0.0414829 ,
       0.09042382, 0.18490181, 0.13457705, 0.18852313, 0.10276868,
       0.13367503, 0.1713421 , 0.21513443, 0.09491988, 0.15915762,
       0.15915762, 0.27444468,        nan, 0.18961997, 0.13457705,
       0.19141571, 0.03853465, 0.02273904, 0.13457705, 0.09491988,
              nan, 0.18499214, 0.11995715, 0.13547365, 0.20891465,
       0.14707062, 0.18756717, 0.14443708, 0.00268779, 0.14707062,
       0.18852313, 0.22890247, 0.14707062, 0.73026008, 0.14707062,
       0.10276868, 0.14443708, 0.14707062, 0.11698202, 0.04683438,
       0.17519231, 0.13457705, 0.09491988, 0.18756717])
```
so when I run `observation['scores'][0][action_set].argmax()` it returns `0`
the variable it chose to branch is wrong?

(3) when I set `expert_probability=1`, the StrongBranching does not have `nan` in `observation['scores'][0][action_set]`

so I'm confused if there's a bug in ecole.observation.Pseudocosts?

**Comment by ghost** [comment]
here's my `instance_0000.lp` used in code (rename the file by removing the suffix `.csv` )
[instance_0000.lp.csv](https://github.com/ds4dm/ecole/files/7902086/instance_0000.lp.csv)

**Comment by AntoinePrv** [comment, marked as answer]
Hello @CXCellTrack,

1. I can reproduce that `observation['scores'][0].shape == 994`. Furthermore, we also have
   `instance.as_pyscipopt().getNVars() == 1000` and `env.model.as_pyscipopt().getNVars() == 994`.
   Somwhere in the preprocessing SCIP must realize that some variables can be eliminated (or even are unused).
   Ecole get the data from SCIP, but it is true that it can sometimes be surprising.
   
2. Yes, we fill in `NaN` when the observation can not be computed for a variable. Here you have `ecole.environment.Branching(..., pseudo_candidates=True)`, so I assume that some `Pseudocosts` can not be computed for pseudo candidates.

3. That must be because `Pseudocosts` cannot be computed for pseudo candidates but `StrongBranchingScores` can.

Hope this helps.

**Reply by ghost** [reply]
thanks for your reply, 
when I use `ecole.environment.Branching(..., pseudo_candidates=False)` the `NaN` disappeared, though I'm not very clear about what does `pseudo_candidates` mean, it works :)

**Reply by AntoinePrv** [reply]
`pseudo_candiadates=True` means the `Branching` environment will return a larger set of branching variables for the action set, extracted with `SCIPgetPseudoBranchCands`.
If I'm correct these are all non fixed variables, even if their value in the solution of the relaxation is integer.

With `pseudo_candidates=False`,  the `Branching` environment will compute the action set with `SCIPgetBranchCands` that only include variables whose solution value to the relaxation is fractional.

**Reply by ghost** [reply]
thanks very much, I think I've got it!
